lib/jruby-kafka/consumer.rb

```diff
@@ -1,41 +1,115 @@
+  #   the form "host:port[path][,host:port[path]*". Required.
+  # @option config [String]  :group_id The consumer group this instance is a member of. Required.
+  # @option config [String]  :topic The topic to consume from.
+  # @option config [String]  :include_topics The inclusive (white) regular expression filter matching
+  #   topics to consume from.
+  # @option config [String]  :exclude_topics The exclusive (black) regular expression filter matching
+  #   topics not to consume from.
+  # @option config [Integer] :num_streams (1) The number of message streams to create.
+  # @option config [String]  :key_decoder ('kafka.serializer.DefaultDecoder') Java class name for
+  #   message key decoder.
+  # @option config [String]  :msg_decoder ('kafka.serializer.DefaultDecoder') Java class name for
+  #   message value decoder.
+  #
+  # One and only one of :topic, :include_topics, or :exclude_topics must be provided.
+  #
+  # For other configuration properties and their default values see 
+  # https://kafka.apache.org/08/configuration.html#consumerconfigs and
+  # https://github.com/apache/kafka/blob/0.8.2.1/core/src/main/scala/kafka/consumer/ConsumerConfig.scala#L90-L182.
+  #
+  def initialize(config={})
+    validate_arguments config
+
+    @properties     =  config.clone
+    @topic          =  @properties.delete :topic
+    @include_topics =  @properties.delete :include_topics
+    @exclude_topics =  @properties.delete :exclude_topics
+    @num_streams    = (@properties.delete(:num_streams) || 1).to_java Java::int
+    @key_decoder    =  @properties.delete(:key_decoder) || 'kafka.serializer.DefaultDecoder'
+    @msg_decoder    =  @properties.delete(:msg_decoder) || 'kafka.serializer.DefaultDecoder'
+
+    @consumer = Java::KafkaConsumer::Consumer.createJavaConsumerConnector create_config
   end
 
-  def run
-    it = @m_stream.iterator
-    begin
-      while it.hasNext
-        begin
-          @m_queue << it.next
-        end
-      end
-    rescue Exception => e
-      puts("#{self.class.name} caught exception: #{e.class.name}")
-      puts(e.message) if e.message != ''
-      if @m_restart_on_exception
-        sleep(@m_sleep_ms)
-        retry
-      else
-        raise e
-      end
+  # Start fetching messages.
+  #
+  # @return [Array<Java::KafkaConsumer::KafkaStream>] list of stream, as specified by the 
+  #   :num_stream configuration parameter. A stream is essentially a queue of incomnig messages
+  #   from Kafka topic partitions.
+  #
+  # @see http://apache.osuosl.org/kafka/0.8.2.1/scaladoc/index.html#kafka.consumer.KafkaStream
+  # 
+  # @note KafkaStreams instances are not thread-safe.
+  def message_streams
+    key_decoder_i = Java::JavaClass.for_name(@key_decoder).
+      constructor('kafka.utils.VerifiableProperties').new_instance nil
+    msg_decoder_i = Java::JavaClass.for_name(@msg_decoder).
+      constructor('kafka.utils.VerifiableProperties').new_instance nil
+
+    if @topic
```

Comment:
There is no way to change `@topic` from outside class, as `@topic` is not an attribute.  You could make use of `attr_reader` and only use the `topic` attribute within the class, but that just seems like wearing suspenders with a belt.  `@topic` is for all intents and purposes already immutable since its not modified after initialization.